rustql-extractor/src/main.rs

```diff
@@ -29,163 +31,198 @@ extern crate syntax;
+            continue;
+        }
+
+        let value = arg.get(1).cloned().or_else(|| args.next());
+        if value.as_ref().map_or(false, |p| pred(p)) {
+            return value;
+        }
+    }
+    None
+}
+
+impl rustc_driver::Callbacks for Callbacks {
+    fn after_analysis(&mut self, compiler: &interface::Compiler) -> Compilation {
+        compiler
+            .global_ctxt()
+            .unwrap()
+            .peek_mut()
+            .enter(|tcx| {
+                let crate_name_env = env::var("CARGO_PKG_NAME").unwrap_or(String::from("main"));
+                let crate_version = (
+                    env::var("CARGO_PKG_VERSION_MAJOR")
+                        .unwrap_or(String::from("0"))
+                        .parse::<u64>()
+                        .unwrap(),
+                    env::var("CARGO_PKG_VERSION_MINOR")
+                        .unwrap_or(String::from("0"))
+                        .parse::<u64>()
+                        .unwrap(),
+                    env::var("CARGO_PKG_VERSION_PATCH")
+                        .unwrap_or(String::from("0"))
+                        .parse::<u64>()
+                        .unwrap(),
+                );
+
+                let crate_name = compiler.crate_name().unwrap().take();
+                if crate_name_env != crate_name {
+                    // happens when the crate name contains a '-', this will then get
+                    // renamed to a '_' to become a rust identifier.
+
+                    //println!("\x1b[31mdifferent names!: {}, {}\x1b[0m", crate_name_env, crate_name);
+                }
+
+                let hir_map = &tcx.hir();
+                let ref krate = hir_map.krate();
+
+                //
+                // assume, crate num of 0 means current crate
+                //
+                let config_hash: String = tcx.crate_hash(hir::def_id::CrateNum::new(0)).to_string();
+                let mut cv = CrateVisitor {
+                    crate_data: data::Crate::new(&crate_name, crate_version, &config_hash),
+                    current_function: None,
+                    // crate_name: crate_name,
+                    map: hir_map,
+                    tcx: tcx,
+                    local_modules: BTreeMap::new(),
+                };
+
+                // add root module
+                // cv.visit_mod(&krate.module, krate.span, CRATE_NODE_ID);
+                cv.crate_data.mods.push(data::Mod {
+                    name: crate_name.to_owned(),
+                    parent_mod: None,
+                });
+
+                walk_crate(&mut cv, krate);
+
+                //println!("{:?}", cv.crate_data);
+                let result = export_crate(&cv.crate_data);
+                if let None = result {
+                    println!(
+                        "ERROR exporting crate: {}",
+                        cv.crate_data.metadata.get_filename()
+                    );
+                }
+            });
+        Compilation::Stop
```

Comment:
What happens when the extractor is also invoked on dependencies? At least my understanding was that if we prematurely stop the compilation while compiling dependencies, then dependent crates cannot be compiled anymore.

Reply:
You are right. I was not aware of that, but it is fixed now.